### PR TITLE
New: Kit defaults - Introduction to the feature [ED-8174]

### PR DESCRIPTION
**Pull request by Nevoss**

### CODE REVIEW

File: modules/kits-elements-defaults/assets/js/commands/create.js [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+	];
+
+	validateArgs( args = {} ) {
+		this.requireContainer();
+	}
+
+	async apply( { container } ) {
+		$e.internal( 'panel/state-loading' );
+
+		const type = container.model.get( 'widgetType' ) || container.model.get( 'elType' );
```
Comment (matipojo) [inline modules/kits-elements-defaults/assets/js/commands/create.js:15]: Try to find an existing method

File: modules/kits-elements-defaults/assets/js/commands/create.js [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+			return {
+				...carry,
+				...extractor( container ),
+			};
+		}, {} );
+
+		try {
+			const createAndRefetch = async () => {
+				await $e.data.create( 'kits-elements-defaults/index', { settings }, { type } );
+
```
Comment (matipojo) [inline modules/kits-elements-defaults/assets/js/commands/create.js:27]: Try `{cache: false}`
```suggestion
			const response = await $e.data.create( 'kits-elements-defaults/index', { settings }, { type } );

```

File: modules/kits-elements-defaults/assets/js/commands/create.js [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+				$e.data.cache.storage.removeItem( 'kits-elements-defaults' );
+
+				await $e.data.get( 'kits-elements-defaults/index' );
+			};
+
+			await Promise.allSettled( [
+				createAndRefetch(),
+				new Promise( ( reslove ) => setTimeout( reslove, 800 ) ),
+			] );
```
Comment (matipojo) [inline modules/kits-elements-defaults/assets/js/commands/create.js:36]: Try to remove

File: modules/kits-elements-defaults/assets/js/e-component.js [hunk 4]
```diff
@@ -0,0 +1,54 @@
+import * as hooks from './hooks';
+import * as dataCommands from './data-commands/';
+import * as commands from './commands/';
+
+export default class EComponent extends $e.modules.ComponentBase {
```
Comment (matipojo) [inline modules/kits-elements-defaults/assets/js/e-component.js:5]:
```suggestion
export default class Component extends $e.modules.ComponentBase {
```

File: modules/kits-elements-defaults/assets/js/editor.js [hunk 5]
```diff
@@ -0,0 +1,6 @@
+import EComponent from './e-component';
```
Comment (matipojo) [inline modules/kits-elements-defaults/assets/js/editor.js:1]: Rename to editor/module.js

File: modules/kits-elements-defaults/assets/js/e-component.js [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+	registerAPI( ...args ) {
+		window.addEventListener( 'elementor/init', async () => {
+			// Load default values on init.
+			$e.data.cache.storage.removeItem( 'kits-elements-defaults' );
+
+			await $e.data.get( 'kits-elements-defaults/index' );
```
Comment (matipojo) [inline modules/kits-elements-defaults/assets/js/e-component.js:11]: Explain

File: modules/kits-elements-defaults/assets/js/e-component.js [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+		} );
+
+		elementor.hooks.addFilter( 'elements/widget/contextMenuGroups', this.addContextMenuItem );
```
Comment (matipojo) [inline modules/kits-elements-defaults/assets/js/e-component.js:14]: Move to module.js

File: modules/kits-elements-defaults/assets/js/e-component.js [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+
+	defaultData() {
+		return this.importCommands( dataCommands );
+	}
+
+	defaultCommands() {
+		return this.importCommands( commands );
+	}
+
+	addContextMenuItem( groups, view ) {
```
Comment (matipojo) [inline modules/kits-elements-defaults/assets/js/e-component.js:35]: Move to module.js

File: modules/kits-elements-defaults/assets/js/hooks/data/fill-defaults-on-drop.js [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+		return 'preview/drop';
+	}
+
+	getId() {
+		return 'fill-defaults-on-drop';
+	}
+
+	getConditions( args ) {
+		return args.model.elType || args.model.widgetType;
+	}
```
Comment (matipojo) [inline modules/kits-elements-defaults/assets/js/hooks/data/fill-defaults-on-drop.js:12]: Try to remove

File: modules/kits-elements-defaults/module.php [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+	public static function get_experimental_data() {
+		return [
+			'name' => 'kits-elements-defaults',
+			'title' => __( 'Kit Elements Defaults', 'elementor' ),
+			'description' => __( 'Set default values for all elements in a kit.', 'elementor' ),
+			'release_status' => Experiments_Manager::RELEASE_STATUS_ALPHA,
+		];
+	}
+
+	public function enqueue_scripts() {
```
Comment (matipojo) [inline modules/kits-elements-defaults/module.php:27]: Private

**Review by matipojo — CHANGES_REQUESTED** [review]

File: modules/kits-elements-defaults/assets/js/commands/create.js (hunk 1 above)
Comment (Nevoss) [reply modules/kits-elements-defaults/assets/js/commands/create.js:15]: Didn't find any

File: modules/kits-elements-defaults/assets/js/commands/create.js (hunk 2 above)
Comment (Nevoss) [reply modules/kits-elements-defaults/assets/js/commands/create.js:27]: Not working

File: modules/kits-elements-defaults/assets/js/commands/create.js (hunk 3 above)
Comment (Nevoss) [reply modules/kits-elements-defaults/assets/js/commands/create.js:36]: Done

File: modules/kits-elements-defaults/assets/js/commands/create.js [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+
+			// Here we should open some toast
+		} catch ( e ) {
+			// Show some error.
```
[the remark this answers is not in the document]
Comment (Nevoss) [reply modules/kits-elements-defaults/assets/js/commands/create.js:40]: Will do it in ED-8180

File: modules/kits-elements-defaults/assets/js/e-component.js (hunk 4 above)
Comment (Nevoss) [reply modules/kits-elements-defaults/assets/js/e-component.js:5]: Done

File: modules/kits-elements-defaults/assets/js/editor.js (hunk 5 above)
Comment (Nevoss) [reply modules/kits-elements-defaults/assets/js/editor.js:1]: Done

File: modules/kits-elements-defaults/assets/js/e-component.js (hunk 6 above)
Comment (Nevoss) [reply modules/kits-elements-defaults/assets/js/e-component.js:11]: Done

File: modules/kits-elements-defaults/assets/js/e-component.js (hunk 8 above)
Comment (Nevoss) [reply modules/kits-elements-defaults/assets/js/e-component.js:35]: Done

File: modules/kits-elements-defaults/module.php (hunk 10 above)
Comment (Nevoss) [reply modules/kits-elements-defaults/module.php:27]: Done

File: modules/kits-elements-defaults/assets/js/editor/commands/create.js [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+		}, {} );
+
+		try {
+			await $e.data.create( 'kits-elements-defaults/index', { settings }, { type } );
+
+			// Refresh Cache.
+			// Cannot use {refresh: true} in `get()` because in the hooks there must be a way to get the data
+			// in sychronous way, and when using `refresh: true`, the data will not be available in syncronous way.
+			$e.data.cache.storage.removeItem( 'kits-elements-defaults' );
+			await $e.data.get( 'kits-elements-defaults/index' );
```
Comment (matipojo) [inline modules/kits-elements-defaults/assets/js/editor/commands/create.js:31]: Extract and use `loadDefaultValues` from the module.

**Review by matipojo — CHANGES_REQUESTED** [review]

File: modules/kits-elements-defaults/assets/js/editor/module.js [hunk 13]
```diff
@@ -0,0 +1,46 @@
+import Component from './component';
+
+( () => {
```
[the remark this answers is not in the document]
Comment (Nevoss) [reply modules/kits-elements-defaults/assets/js/editor/module.js:3]: Why do I need this module?

File: modules/kits-elements-defaults/assets/js/editor/commands/create.js (hunk 12 above)
Comment (Nevoss) [reply modules/kits-elements-defaults/assets/js/editor/commands/create.js:31]: Cannot extract from module + feels like creating method to wrap a command is too much wrapping

Comment (matipojo) [reply modules/kits-elements-defaults/assets/js/editor/commands/create.js:31]: You already have a `loadDefaultValues()` function on:
https://github.com/elementor/elementor/pull/19778/files#diff-4715101938179d6613bb7a4945c287b4b4715602452c587dddeb3120e92c51afR9-R10

File: modules/kits-elements-defaults/assets/js/editor/module.js (hunk 13 above)
[the remark this answers is not in the document]
Comment (matipojo) [reply modules/kits-elements-defaults/assets/js/editor/module.js:3]: For consistency

**Review by matipojo — CHANGES_REQUESTED** [review]
Please cover with Jest tests

File: modules/kits-elements-defaults/assets/js/editor/module.js (hunk 13 above)
[the remark this answers is not in the document]
Comment (Nevoss) [reply modules/kits-elements-defaults/assets/js/editor/module.js:3]: Done

File: modules/kits-elements-defaults/assets/js/editor/commands/create.js (hunk 12 above)
Comment (Nevoss) [reply modules/kits-elements-defaults/assets/js/editor/commands/create.js:31]: Done

File: modules/kits-elements-defaults/assets/js/editor/module.js [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+		// in sychronous way, and when using `refresh: true`, the data will not be available in syncronous way.
+		$e.data.cache.storage.removeItem( 'kits-elements-defaults' );
+		$e.data.get( 'kits-elements-defaults/index' );
+	}
+
+	function addContextMenuItem() {
+		elementor.hooks.addFilter( 'elements/widget/contextMenuGroups', ( groups, view ) => {
+			if ( ! elementor.config?.user?.is_administrator ) {
+				return groups;
+			}
```
[the remark this answers is not in the document]
Comment (Nevoss) [reply modules/kits-elements-defaults/assets/js/editor/module.js:17]: Done

**Comment by Nevoss** [comment]
@matipojo This is only the skeleton, tests will be in each task

**Review by matipojo — CHANGES_REQUESTED** [review]
Please fix the checks errors

**Pull request merged**